_drafts/reading-legacy-code-to-harmonize-ruby-and-rails-code-loaders.md

```diff
@@ -0,0 +1,701 @@
+
+Let's retry the reproduction steps:
+
+{% highlight sh %}
+cd ~/git/rapid7/pro/engine
+RAILS_ENV=development ./msfpro
+{% endhighlight %}
+
+{% highlight irb %}
+msf > irb
+>> Object.send(:remove_const, :SocialEngineering)
+>> exit
+msf > use auxiliary/pro/social_engineering/web_phish
+msf > set CAMPAIGN_ID 1
+msf > run
+{% endhighlight %}
+
+![debugging output](/images/debugging-output.png)
+
+Ok, so it's a namespace Module for `Metasploit3` and `Metasploit3` happens to be the name of the class in `auxiliary/pro/social_engineering/web_phish`.
+
+![`auxiliary/pro/social_engineering/web_phish` `Metasploit3` class](/images/auxiliary-pro-social-engineering-web-phish-class.png)
+
+The `MetasploitN` class naming convention is used to indicate the minimum major version of Metasploit Framework required to run the module, so there are 100s of `Metasploit3` and `Metasploit4` classes.
+
+### Tracking anonymous Module creation
+
+So, however `use` loads the Metasploit Module, it must introduce an anonymous Module (`Module.new`) that is used to prevent the `Metasploit3` classes from colliding with each other.  So, let's look for `Module.new` in `metasploit-framework/lib`.
+
+![`Module.new` in `metasploit-framework/lib`](/images/module-new-in-metasploit-framework-lib.png)
+
+If I was declaring an anonymous `Module` to wrap something, `wrap` would be a good name, so let's look at each of those.
+
+![`Module.new` in `Msf::ModuleManager#reload_module`](/images/module-new-in-msf-module-manager-reload-module.png)
+
+The first one is in `Msf::ModuleManager#reload_module`.  It is unlikely that the only way to load modules is to use a function call `reload`, so moving on...
+
+![`Module.new` in `Msf::ModuleManager#load_module_from_file`](/images/module-new-in-msf-module-manager-load-module-from-file.png)
+
+`Msf::ModuleManager#load_module_from_file` is probably the correct method, but let's check the last line.
+
+![`Module.new` in `Msf::ModuleManager#load_module_from_archive`](/images/module-new-in-msf-module-manager-load-module-from-archive.png)
+
+Huh, so `Msf::ModuleManager#load_module_from_file` and `Msf::ModuleManager#load_module_from_archive` look really similar, except one using `Msf::ModuleManager#load_module_source` and the other is using `Fastlib.load`.  We'll skip over what `Fastlib` is for now and stick with the `Module.new` in `Msf::ModuleManager#load_module_from_path`.  We need to read all of `Msf::ModuleManager#load_module_from_file` to understand the context of the `Module.new` call on line 864.
+
+### De-anonymizing namespace Modules
+
+The minimal change I can make to `Msf::ModuleManager#load_module_from_file` is that I replace the direct call to `Module.new` with a method that returns an named Module instead.  Here's the constraint this new method must fulfill from the code we've seen:
+
+1. The name must be unique, so that it prevent naming collision
```

Comment:
1. The name must be unique, so that it **~~prevent~~** **prevents** naming collision